### DISCUSSION TITLE
How can I use `py5_tools.screenshot` in class mode?

### DISCUSSION THREAD

**Discussion by zztkm** [py5coding/py5generator#170] [Q&A] [answered]
I'm currently trying to use `py5_tools.screenshot` in class mode, but I'm having trouble getting it to work, so any advice would be great!

```python
from py5 import Sketch
from py5_tools import screenshot

from utils import get_timestamp


class FirstSketch(Sketch):
    def settings(self):
        self.size(600, 600)

    def setup(self):
        self.rect_mode(self.CENTER)

    def draw(self):
        self.rect(self.mouse_x, self.mouse_y, 30, 30)

    def key_typed(self):
        if self.key == "s":
            time_stamp = get_timestamp()
            img = screenshot(sketch=self, hook_post_draw=True)
            img.save(f"{time_stamp}.png")


def main():
    sketch = FirstSketch()
    sketch.run_sketch()


if __name__ == "__main__":
    main()
```

When typing `s`,  got error output:
```shell
❯ python main.py
py5 encountered an error in your code:File "C:\Users\zztkm\dev\sandbox\pydev\py5-example\main.py", line 20, in key_typed
    17   def key_typed(self):
    18       if self.key == "s":
    19           time_stamp = get_timestamp()
--> 20           img = screenshot(sketch=self, hook_post_draw=True)
    21           img.save(f"{time_stamp}.png")
    ..................................................
     self = <__main__.FirstSketch object at 0x000002882B98BA90>
    ..................................................

UnboundLocalError: local variable 'py5' referenced before assignment
```

**Comment by hx2A** [comment, marked as answer]
Hi, @zztkm! It seems like you've found a bug. Thank you for reporting this.

There are few things you can do differently. First, I recommend you use `save_frame()` instead of `py5_tools.screenshot()` here.

https://py5coding.org/reference/sketch_save_frame.html

I created `py5_tools.screenshot()` with the goal of providing an easy way for you to grab an arbitrary screenshot of a running sketch and have the image embedded in a Jupyter notebook. I should make it more clear in the documentation for `py5_tools.screenshot()` that `save_frame()` is also available and is a better choice for saving a frame from within a running Sketch.

Still, the code you wrote should work if `hook_post_draw=False`. The `hook_post_draw` is used if your Sketch has a `post_draw()` method, which is only relevant if you are using a Processing library that calls it like Camera3D.

Another solution is to remove `hook_post_draw` (it defaults to `False`) and add an `import py5` to the other imports. You shouldn't have to add that import though. This bug will be fixed in the next release.

**Reply by hx2A** [reply]
@zztkm , I made a mistake in my last message. 

> Still, the code you wrote should work if `hook_post_draw=False`

I just realized this is incorrect. I made a coding change for #99 so that `py5_tools.screenshot()` throws an error if you try to call it from within a Sketch as you were doing in your example. It isn't designed to work that way. While making that coding change, I also introduced the error that you are getting here. I do need to fix that, but once that is fixed, you'll get the error message added for #99.

Bottom line, you should use the `save_frame()` method to save the frame from your `key_typed()` function.

**Comment by zztkm** [comment]
Hi @hx2A! Thanks for the reply!

Oh, it was a bug. I thought I had written some crazy code!

There was one mistake in the code I posted.
I accidentally set `hook_post_draw=True` while debugging, but when I set it to False and ran it again, the same error occurred.

supplement:
I use py5==0.8.2a1

**Comment by zztkm** [comment]
@hx2A 

I took your advice and used `save_frame` and got the image as expected! Thank you!

Here is the code that worked correctly in case others encounter similar problems.

```python
from datetime import datetime

from py5 import Sketch


def get_timestamp() -> str:
    now = datetime.now()
    return now.strftime("%Y-%m-%d-%H-%M-%S")


class FirstSketch(Sketch):
    def settings(self):
        self.size(600, 600)

    def setup(self):
        self.rect_mode(self.CENTER)

    def draw(self):
        self.rect(self.mouse_x, self.mouse_y, 30, 30)

    def key_typed(self):
        if self.key == "s":
            time_stamp = get_timestamp()
            self.save_frame(f"{time_stamp}.png")
            print(f"save to {time_stamp}.png")


def main():
    sketch = FirstSketch()
    sketch.run_sketch()


if __name__ == "__main__":
    main()
```

output:
```shell
> python main.py
save to 2022-10-04-09-40-56.png
```

2022-10-04-09-40-56.png
![2022-10-04-09-40-56](https://user-images.githubusercontent.com/33755694/193710420-4734dcf2-7943-4563-b6ff-810b83502e65.png)

**Reply by hx2A** [reply]
This new code is perfect. Good job, @zztkm !

You did uncover a bug here. Instead of getting the error that you received, "UnboundLocalError: local variable 'py5' referenced before assignment", you should have gotten the error message "Calling py5_tools.screenshot() from within a py5 user function is not allowed. Please move this code to outside the Sketch." I'm going to improve that error message and add to that, " or consider using save_frame() instead."

**Reply by zztkm** [reply]
That error message is very user friendly!

Can I improve it too?
I hope I can contribute in some small way with the code, such as improving the error message!

**Reply by hx2A** [reply]
@zztkm , would you like to make a PR for this? I can walk you through the steps you'll need to take to make a PR. It's straightforward, you'll need to move a new lines of code around and edit a string containing the error message.

**Reply by zztkm** [reply]
@hx2A , I'll try it!

**Reply by zztkm** [reply]
@hx2A Error messages have been corrected!

#172
